Adds support for managing a separate locale for dates

dzena opened the pull request:
This PR includes minimal changes to the API. No breaking changes introduced. If more sophisticated solution is needed, I'll be happy to redo whats needed.

Commit:
[WIP] add 'dateLanguage'

Commit:
feat - add 'dateLanguage' to support multiple locale at the same time

Commit:
chore - update README.md

Commit:
chore - update README.md - rm unnecessary comments

robisim74 commented on README.md:
````diff
@@ -326,6 +326,46 @@ Angular l10n types that it is useful to know:
+Same as _language_ property, there is a _dateLanguage_ property. It gives us a possibility to use a separate locale (other than the one used for translations) to format dates.
+#### Pure Pipes
+```Html
+<p>{{ today | l10nDate:locale.dateLanguage:{ dateStyle: 'full', timeStyle: 'short' } }}</p>
+```
+#### Directives
+```Html
+<p [options]="{ dateStyle: 'full', timeStyle: 'short' }" l10nDate>{{ today }}</p>
+```
+
+If you set a _dateLanguage_, the directives will use it by default (it will fallback to _language_ id not set). There is a possibility to pass a prop to the directive if you want to use _language_ instead of _dateLanguage_.
+
+```Html
+<p [useDateLanguage]="false" [options]="{ dateStyle: 'full', timeStyle: 'short' }" l10nDate>{{ today }}</p>
+```
+
+#### APIs
+```TypeScript
+export class AppComponent implements OnInit {
+
+    constructor(private translation: L10nTranslationService, private intl: L10nIntlService) { }
+
+    ngOnInit() {
+        this.translation.onChange().subscribe({
+            next: () => {
+                this.formattedToday = this.intl.formatDate(this.today, { dateStyle: 'full', timeStyle: 'short' });
+            }
+        });
+    }
+
+}
+```
+
+To use _language_ over _dateLanguage_ just pass it to the `formatDate` method (note the last param - `false`).
+
+```TypeScript
+this.intl.formatDate(this.today, { dateStyle: 'full', timeStyle: 'short' }, false);
+     
+```
+
````
The use of this feature should not be under `Intl Api`.

dzena replied:
I'll remove all my changes from the README.

robisim74 commented on projects/angular-l10n-app/src/app/app.component.html:
```diff
@@ -10,6 +10,7 @@
 
-<router-outlet></router-outlet>
\ No newline at end of file
+<router-outlet></router-outlet>
```
The sample app should contain only the basic functionality, to offer developers a starter template.

dzena replied:
Makes sense. Understand.

robisim74 commented on projects/angular-l10n/src/lib/directives/l10n-date.directive.ts:
```diff
@@ -14,6 +14,7 @@ export class L10nDateDirective extends L10nDirective {
     }
 
     @Input() public options: any;
+    @Input() public useDateLanguage = true;
```
The idea is interesting, but the default value should be false, since it is an option.

dzena replied:
I will remove it.

robisim74 commented on projects/angular-l10n/src/lib/directives/l10n-date.directive.ts:
```diff
@@ -25,7 +26,7 @@ export class L10nDateDirective extends L10nDirective {
 
     protected getValue(text: string): string {
-        return this.intl.formatDate(text, this.options);
+        return this.intl.formatDate(text, this.options, this.useDateLanguage);
```
The `formatDate` method of `L10nIntlService` should not be changed, it should be the responsibility of the caller to pass the desired locale. I am thinking that perhaps it is not even necessary to change the directives, as they can be easily customized by extending the basic `L10nDirective`.

dzena replied:
So, the reason behind this is following.
If you use English for translations and German for dates and you have a date directive that is displaying a date with `[options]="{dateStyle: 'full'}"` you end up with date labels (day and month) displayed in German and the rest of the web app is in English. Hence, I wanted a way to decide which locale to use for the dates.
First I decided to just use `@Input() language` in the directive, but then we might have the following problem.
The base `L10nDirective` is listening for the translation changes and the user is passing an input with the language, `<p [language]="locale.dateLanguage" [options]="{ dateStyle: 'full' }" l10nDate>{{ today }}</p>`. If the `L10nDirective` is faster, then the wrong language will be used to format the date.
What would be your approach here?

robisim74 replied:
Adding the language property to directives (and also to validation directives and async pipes) would be great, because all components would be aligned and would allow you to manage multilingual components!

So something like this:
```
export class L10nDateDirective extends L10nDirective {
    [...]

    @Input() public language: string;

    [...]
    protected getValue(text: string): string {
        return this.intl.formatDate(text, this.options, this.language || undefined);
    }
}
```
should work, because:

case 1) If the `ngOnChanges` method of the `L10nDirective` https://github.com/robisim74/angular-l10n/blob/a42ef6499300cb85b6674d3a6ddab63db4f4e7bc/projects/angular-l10n/src/lib/models/l10n-directive.ts#L39 is fired before the `addTranslationListener` method, https://github.com/robisim74/angular-l10n/blob/a42ef6499300cb85b6674d3a6ddab63db4f4e7bc/projects/angular-l10n/src/lib/models/l10n-directive.ts#L112 `getValue` will update using the new language

case 2) If the `ngOnChanges` method is fired after, `getValue `will use the new language

Ok, it is fired twice! We should find a way not to turn it on twice.

dzena replied:
Will this mean that the language property needs to be mandatory?
In this case, we can get rid of the `addTranslationListener` and just rely on `ngOnChanges` I guess. This will also help with the problem `getValue` being called twice.

But this seems like a really nasty breaking change. The developer will need to pass the language to all directives.

robisim74 replied:
No, the language property would not be mandatory. If one decides to pass the new language, that would be used, otherwise everything would continue to work as before.

dzena replied:
What do you think about this solution there (prevents calling the `getValue` twice)?

Inside the base `L10nDirective`.

```
@Input() language: string;

public ngAfterViewInit(): void {
        if (this.el && this.el.nativeElement) {
            this.element = this.el.nativeElement;
            this.renderNode = getTargetNode(this.element);
            this.text = this.getText();
            this.attributes = this.getAttributes();
            this.addTextListener();

           // THIS IS CHANGED
            if ( this.language ) {
                this.replaceText()
                this.replaceAttributes();
            } else {
                this.addTranslationListener();
            }
        }
    }

```
We don't need to register the translation listener if the language prop exists. `ngOnChanges` will take care of `replaceTexet()` and `replaceAttributes`?

robisim74 replied:
Yes, it should work properly, because `ngOnChanges `activates when properties change value (and it calls the replace methods)

robisim74 commented on projects/angular-l10n/src/lib/models/types.ts:
```diff
@@ -25,6 +25,15 @@ export interface L10nLocale {
+     * - region: ISO 3166 two-letter, uppercase code
+     * - extension: 'u' (Unicode) extensions
+     */
+    dateLanguage?: string;
```
Perfect. We have to provide an appropriate interface, which allows the user to customize the use of the library.

robisim74 commented on projects/angular-l10n/src/lib/services/l10n-intl.service.ts:
```diff
@@ -28,13 +28,15 @@ import { L10nTranslationService } from './l10n-translation.service';
         value: any,
         options?: L10nDateTimeFormatOptions,
-        language = this.locale.language,
+        useDateLanguage = true,
+        language = useDateLanguage ? this.locale.dateLanguage || this.locale.language : this.locale.language,
```
This way you are changing the API, creating a breaking change. It wouldn't be a problem, but I don't like boolean parameters: it's the caller's responsibility to provide the desired data, without having to change the method signature.

dzena replied:
Ah, yes, you are right. I'll change it.
Would you try to fallback to `dateLanguage` first and than to `language`?

robisim74 replied:
We should leave it as it was:
```
    public formatDate(
        value: any,
        options?: L10nDateTimeFormatOptions,
        language = this.locale.language,
        timeZone = this.locale.timeZone
    )
```

robisim74 replied:
@dzena Maybe you were right here, it is preferable to fallback to `dateLanguage `first and than to `language`:

```
    public formatDate(
        value: any,
        options?: L10nDateTimeFormatOptions,
        language = this.locale.dateLanguage || this.locale.language,
        timeZone = this.locale.timeZone
    )
```

So if `dateLanguage `has a value, it will be used by default, even without evaluating the input parameter.

dzena replied:
@robisim74 
yes, i think it should be like that. So, if i've set the `dateLanguage` I would expect that it is used by default (no need to pass a `language`). 
I see that you did it already. Thanks a lot!
Cheers

robisim74 reviewed (changes requested):
@dzena Thanks for the excellent contribution and effort. I have therefore tried to carefully review your every modification proposal point by point.

Basically:
- I ask you to avoid updating the Readme and the sample app until the new feature is released (if I merge the doc of a feature not yet released it would create confusion)
- I suggest you make this feature simpler and more dynamic, in the style of the new version of the library

Let me know.

Commit:
chore - simplify the usage of the "dateLanguage"

Commit:
chore - simplify the usage of the dateLanguage

dzena commented:
@robisim74 
I simplified the usage of `dateLanguage, please give it another try.
Also, I have removed all unnecessary code (from the README and the sample app).

robisim74 commented:
Looks good! Thanks Milan for the effort!
